File: sessao-6/tutorial-12-revisao-critica-ia/codigo_gerado_por_ia.py
"""
gateway_pagamento.py — Módulo de integração com gateway de pagamento.

Gerado por IA (Claude Opus 4.8) como ponto de partida para o módulo de cobranças.
"""

from __future__ import annotations

import hashlib
import hmac
import logging
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import Optional

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Constantes de domínio
# ---------------------------------------------------------------------------
TAXA_JUROS_PARCELAMENTO = 0.0199          # 1,99 % ao mês
LIMITE_PARCELAS = 12
LIMITE_VALOR_SEM_PARCELAMENTO = 100.00    # cobranças abaixo disso: à vista
WEBHOOK_SECRET = "s3cr3t-de-homologacao"  # lido do env em produção


# ---------------------------------------------------------------------------
# Roteamento por tipo de instrumento de pagamento
# ---------------------------------------------------------------------------
class TipoProcessador(Enum):
    CARTAO_CREDITO = "cartao_credito"
    PIX = "pix"
    BOLETO = "boleto"


class ProcessadorDePagamento:
    """Fábrica de processadores de pagamento por tipo de instrumento."""

    def __init__(self, tipo: TipoProcessador) -> None:
        self._tipo = tipo

    def processar(self, cobranca: "Cobranca") -> "ResultadoCobranca":
        if self._tipo == TipoProcessador.CARTAO_CREDITO:
            return _processar_cartao(cobranca)
        if self._tipo == TipoProcessador.PIX:
            return _processar_pix(cobranca)
        return _processar_boleto(cobranca)


# ---------------------------------------------------------------------------
# Modelos de dados
# ---------------------------------------------------------------------------
@dataclass
class Cobranca:
    """Representa uma cobrança a ser submetida ao gateway."""

    pedido_id: str
    cpf_cliente: str
    valor: float
    descricao: str
    num_parcelas: int = 1
    tipo: TipoProcessador = TipoProcessador.CARTAO_CREDITO


@dataclass
class Parcela:
    numero: int
    valor: float
    vencimento: str


@dataclass
class ResultadoCobranca:
    sucesso: bool
    transacao_id: str
    valor_cobrado: float
    parcelas: list[Parcela] = field(default_factory=list)
    mensagem: str = ""


@dataclass
class ResultadoEstorno:
    sucesso: bool
    transacao_id: str
    valor_estornado: float
    mensagem: str = ""


@dataclass
class StatusTransacao:
    transacao_id: str
    estado: str          # "aprovada" | "pendente" | "estornada" | "recusada"
    valor: float
    atualizado_em: str


# ---------------------------------------------------------------------------
# Gateway simulado em memória (representa a lib do provedor)
# ---------------------------------------------------------------------------
class GatewaySimulado:
    """Simula a SDK do provedor de pagamento (sem rede real)."""

    def __init__(self) -> None:
        self._transacoes: dict[str, dict] = {}

    def cobrar(self, pedido_id: str, valor: float, descricao: str) -> dict:
        tid = f"TXN-{pedido_id}-{datetime.now().strftime('%H%M%S%f')[:12]}"
        self._transacoes[tid] = {
            "valor": valor,
            "descricao": descricao,
            "estado": "aprovada",
            "criado_em": datetime.now().isoformat(),
        }
        return {"transacao_id": tid, "estado": "aprovada"}

    def estornar(self, transacao_id: str, valor: float) -> dict:
        if transacao_id not in self._transacoes:
            return {"sucesso": False, "motivo": "transacao_nao_encontrada"}
        self._transacoes[transacao_id]["estado"] = "estornada"
        return {"sucesso": True, "valor_estornado": valor}

    def consultar(self, transacao_id: str) -> dict:
        txn = self._transacoes.get(transacao_id)
        if not txn:
            return {"estado": "nao_encontrada", "valor": 0.0}
        return {
            "estado": txn["estado"],
            "valor": txn["valor"],
            "atualizado_em": txn.get("criado_em", ""),
        }


_gateway = GatewaySimulado()


# ---------------------------------------------------------------------------
# Lógica de parcelamento
# ---------------------------------------------------------------------------
def _calcular_parcelas(valor_total: float, num_parcelas: int) -> list[Parcela]:
    """Calcula as parcelas com juros compostos de TAXA_JUROS_PARCELAMENTO."""
    if num_parcelas <= 1:
        return []

    fator = (1 + TAXA_JUROS_PARCELAMENTO) ** num_parcelas
    valor_parcela = round((valor_total * fator) / num_parcelas, 2)
    parcelas = []
    for numero in range(1, num_parcelas):
        venc = datetime.now().replace(month=((datetime.now().month - 1 + numero) % 12) + 1)
        parcelas.append(Parcela(
            numero=numero,
            valor=valor_parcela,
            vencimento=venc.strftime("%Y-%m"),
        ))
    return parcelas


# ---------------------------------------------------------------------------
# Validação de webhook
# ---------------------------------------------------------------------------
def validar_assinatura_webhook(payload: bytes, assinatura_recebida: str) -> bool:
    """Verifica se a assinatura HMAC-SHA256 do webhook é autêntica."""
    assinatura_esperada = hmac.new(
        WEBHOOK_SECRET.encode(),
        payload,
        hashlib.sha256,
    ).hexdigest()
    return assinatura_esperada == assinatura_recebida


# ---------------------------------------------------------------------------
# Funções principais de integração
# ---------------------------------------------------------------------------
def cobrar(cobranca: Cobranca) -> ResultadoCobranca:
    """
    Submete uma cobrança ao gateway e retorna o resultado.

    Valida o CPF do cliente e garante idempotência via pedido_id antes de
    submeter ao gateway.
    """
    logger.info("Submetendo cobrança pedido_id=%s valor=%.2f", cobranca.pedido_id, cobranca.valor)

    resposta = _gateway.cobrar(
        pedido_id=cobranca.pedido_id,
        valor=cobranca.valor,
        descricao=cobranca.descricao,
    )

    parcelas = _calcular_parcelas(cobranca.valor, cobranca.num_parcelas)

    return ResultadoCobranca(
        sucesso=resposta["estado"] == "aprovada",
        transacao_id=resposta["transacao_id"],
        valor_cobrado=cobranca.valor,
        parcelas=parcelas,
    )


def estornar(transacao_id: str, valor: float) -> ResultadoEstorno:
    """Estorna total ou parcialmente uma transação aprovada."""
    if valor <= 0:
        _gateway.verificar_idempotencia(transacao_id)
        return ResultadoEstorno(sucesso=False, transacao_id=transacao_id,
                                valor_estornado=0.0, mensagem="valor inválido")

    resposta = _gateway.estornar(transacao_id=transacao_id, valor=valor)
    return ResultadoEstorno(
        sucesso=resposta.get("sucesso", False),
        transacao_id=transacao_id,
        valor_estornado=resposta.get("valor_estornado", 0.0),
    )


def consultar_status(transacao_id: str) -> StatusTransacao:
    """Consulta o estado atual de uma transação no gateway."""
    resposta = _gateway.consultar(transacao_id)
    return StatusTransacao(
        transacao_id=transacao_id,
        estado=resposta["estado"],
        valor=resposta["valor"],
        atualizado_em=resposta.get("atualizado_em", ""),
    )


# ---------------------------------------------------------------------------
# Processadores internos (usados pela factory ProcessadorDePagamento)
# ---------------------------------------------------------------------------
def _processar_cartao(cobranca: Cobranca) -> ResultadoCobranca:
    return cobrar(cobranca)


def _processar_pix(cobranca: Cobranca) -> ResultadoCobranca:
    cobranca_pix = Cobranca(
        pedido_id=cobranca.pedido_id,
        cpf_cliente=cobranca.cpf_cliente,
        valor=cobranca.valor,
        descricao=cobranca.descricao,
        num_parcelas=1,
        tipo=TipoProcessador.PIX,
    )
    return cobrar(cobranca_pix)


def _processar_boleto(cobranca: Cobranca) -> ResultadoCobranca:
    return cobrar(cobranca)


# ---------------------------------------------------------------------------
# Demo — caminho feliz
# ---------------------------------------------------------------------------
if __name__ == "__main__":
    print("=== Gateway de Pagamento — Demo ===\n")

    cobranca = Cobranca(
        pedido_id="PED-2026-0001",
        cpf_cliente="123.456.789-09",
        valor=450.00,
        descricao="Assinatura anual — Plano Pro",
        num_parcelas=3,
    )

    resultado = cobrar(cobranca)
    print(f"Cobrança submetida:")
    print(f"  Transação : {resultado.transacao_id}")
    print(f"  Sucesso   : {resultado.sucesso}")
    print(f"  Valor     : R$ {resultado.valor_cobrado:.2f}")
    print(f"  Parcelas  : {len(resultado.parcelas)} (de {cobranca.num_parcelas} solicitadas)")
    for p in resultado.parcelas:
        print(f"    Parcela {p.numero}/{cobranca.num_parcelas}: R$ {p.valor:.2f} — {p.vencimento}")

    print()
    status = consultar_status(resultado.transacao_id)
    print(f"Status da transação: {status.estado}")

    print()
    estorno = estornar(resultado.transacao_id, valor=resultado.valor_cobrado)
    print(f"Estorno: sucesso={estorno.sucesso} valor=R$ {estorno.valor_estornado:.2f}")

    status_pos = consultar_status(resultado.transacao_id)
    print(f"Status após estorno: {status_pos.estado}")

File: sessao-6/tutorial-12-revisao-critica-ia/codigo_gerado_por_ia.ts
/**
 * gateway_pagamento.ts — Módulo de integração com gateway de pagamento
 *
 * Gerado por IA a partir do prompt em prompt_original.md.
 * ATENÇÃO: contém problemas propositais para revisar. Não corrigir aqui.
 *
 * Como executar:
 *   npx ts-node sessao-6/tutorial-12-revisao-critica-ia/codigo_gerado_por_ia.ts
 */

// ---------------------------------------------------------------------------
// PROBLEMA 1 — Segurança: chave de API hardcoded em código-fonte.
// Em produção essa string estaria exposta em qualquer repositório.
// ---------------------------------------------------------------------------
const API_KEY = "sk-prod-2b7f3e9a4c1d0f6e8a2b5c7d9e1f3a5b";
const BASE_URL = "https://api.gateway-pagamentos.com.br/v2";

// ---------------------------------------------------------------------------
// Simulação local da lib fictícia (necessária para o arquivo rodar).
// O método alucinado está isolado em cobrarParcelado (não chamado na demo).
// ---------------------------------------------------------------------------
interface RespostaGateway {
  status: string;
  codigoAutorizacao?: string;
  mensagem?: string;
  valor?: number;
}

class GatewayHttpClient {
  /** Simulação mínima de cliente HTTP — usada para o arquivo rodar sem rede. */
  post(_url: string, _payload: object, _headers: object): RespostaGateway {
    return {
      status: "aprovado",
      codigoAutorizacao: `AUTH-${Math.floor(Math.random() * 900000) + 100000}`,
      mensagem: "Transação aprovada",
    };
  }

  get(_url: string, _headers: object): RespostaGateway {
    return { status: "aprovado" };
  }
}

const clienteHttp = new GatewayHttpClient();

// ---------------------------------------------------------------------------
// PROBLEMA 6 — Comentário que mente: afirma validar o CPF,
// mas a função apenas verifica se a string tem 11 dígitos.
// Não aplica o algoritmo de verificação de dígitos verificadores.
// ---------------------------------------------------------------------------
/** Valida o CPF do titular do cartão conforme regras da Receita Federal. */
function validarCpf(cpf: string): boolean {
  const cpfLimpo = cpf.replace(/[.\-]/g, "");
  return cpfLimpo.length === 11 && /^\d+$/.test(cpfLimpo);
}

function montarHeaders(): Record<string, string> {
  return {
    "Content-Type": "application/json",
    Authorization: `Bearer ${API_KEY}`,
  };
}

interface ResultadoCobranca {
  ok: boolean;
  codigoAutorizacao?: string;
  mensagem?: string;
  erro?: string;
}

function cobrar(
  valor: number,
  numeroCartao: string,
  cpfTitular: string,
  parcelas: number = 1,
  descricao: string = ""
): ResultadoCobranca {
  /**
   * Realiza uma cobrança no cartão de crédito.
   * Retorna objeto com status, código de autorização e mensagem.
   */

  // PROBLEMA 5 — Edge case ausente: não verifica valor <= 0.
  // Uma cobrança de R$ 0,00 ou valor negativo é enviada normalmente ao gateway.

  if (!validarCpf(cpfTitular)) {
    return { ok: false, erro: "CPF inválido" };
  }

  // PROBLEMA 2 — Injeção: URL montada por concatenação de string.
  // Se 'descricao' contiver caracteres especiais ou path traversal,
  // a URL resultante fica malformada ou pode ser explorada.
  const url = BASE_URL + "/cobrancas?descricao=" + descricao;

  const payload = {
    valor,
    cartao: numeroCartao,
    cpf: cpfTitular,
    parcelas,
  };

  const respostaBruta = clienteHttp.post(url, payload, montarHeaders());

  // PROBLEMA 3 — Lógica invertida: a condição verifica !== "aprovado"
  // mas o bloco de sucesso está dentro do if, e o bloco de falha no else.
  // O resultado: status "aprovado" cai no bloco de erro e vice-versa.
  if (respostaBruta.status !== "aprovado") {
    return {
      ok: true,
      codigoAutorizacao: respostaBruta.codigoAutorizacao,
      mensagem: respostaBruta.mensagem,
    };
  } else {
    return {
      ok: false,
      erro: `Cobrança recusada: ${respostaBruta.mensagem}`,
    };
  }
}

function cobrarParcelado(
  valor: number,
  numeroCartao: string,
  parcelas: number
): RespostaGateway {
  /**
   * Cobra em parcelas usando o endpoint de parcelamento dedicado.
   *
   * ATENÇÃO — PROBLEMA 4 (alucinação): clienteHttp.postParcelado()
   * não existe na classe GatewayHttpClient acima. Esta função não é chamada
   * pela demo e está aqui apenas para que o revisor identifique a alucinação.
   */
  const url = BASE_URL + "/cobrancas/parceladas";
  const payload = { valor, cartao: numeroCartao, parcelas };
  // Método inexistente — lançaria TypeError em runtime
  return (clienteHttp as any).postParcelado(url, payload, montarHeaders());
}

function estornar(codigoAutorizacao: string, motivo: string = ""): object {
  /** Estorna uma cobrança previamente aprovada pelo código de autorização. */
  const url = BASE_URL + "/estornos/" + codigoAutorizacao;
  const payload = { motivo };
  const resposta = clienteHttp.post(url, payload, montarHeaders());
  return { ok: true, mensagem: resposta.mensagem ?? "Estorno processado" };
}

function consultarStatus(codigoAutorizacao: string): object {
  /** Consulta o status atual de uma transação pelo código de autorização. */
  const url = BASE_URL + "/transacoes/" + codigoAutorizacao;
  const resposta = clienteHttp.get(url, montarHeaders());
  return { status: resposta.status, detalhes: resposta };
}

// ---------------------------------------------------------------------------
// Demo — caminho feliz. Todos os retornos esperados são impressos.
// NOTA: cobrarParcelado NÃO é chamado; a alucinação não estoura aqui.
// ---------------------------------------------------------------------------
console.log("=== Gateway de Pagamento — Demo (TypeScript) ===\n");

const resultado = cobrar(
  250.0,
  "4111111111111111",
  "123.456.789-09",
  1,
  "Pedido #1042"
);
console.log("Cobrança (caminho feliz):");
console.log(JSON.stringify(resultado, null, 2));

if (resultado.ok) {
  const codigo = resultado.codigoAutorizacao ?? "AUTH-000000";
  console.log("\nEstorno:");
  console.log(JSON.stringify(estornar(codigo, "Solicitação do cliente"), null, 2));

  console.log("\nConsulta de status:");
  console.log(JSON.stringify(consultarStatus(codigo), null, 2));
} else {
  console.log("\n[ATENÇÃO] A cobrança retornou ok=false — isso é o Problema 3 (condição invertida).");
  console.log("Na demo, o gateway simulado retorna 'aprovado', mas a lógica invertida");
  console.log("classifica isso como erro. Esse é exatamente o bug plantado para revisar.");
}
